File: ios/RNBidmadCommonModule.m
_Pragma("clang diagnostic push")
_Pragma("clang diagnostic ignored \"-Wstrict-prototypes\"")

#import <React/RCTBridgeModule.h>
#import <OpenBiddingHelper/OpenBiddingHelper.h>

@interface RCT_EXTERN_MODULE(RNBidmadCommonModule, NSObject)

_RCT_EXTERN_REMAP_METHOD(initializeSdk, initializeSdkWithAppKey:(NSString *)appKey resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(reqAdTrackingAuthorization, reqAdTrackingAuthorizationWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setAdvertiserTracking, setAdvertiserTrackingWithEnable:(BOOL)enable resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(advertiserTracking, getAdvertiserTrackingWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setIsChildDirectedAds, setWithIsChildDirectedAds:(BOOL)isChildDirectedAds resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(isChildDirectedTreatment, isChildDirectedTreatmentWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setUserConsentCCPA, setWithUserConsentStatusForCCPACompliance:(BOOL)userConsentStatusForCCPACompliance resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(isUserConsentCCPA, isUserConsentCCPAWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setIsDebug, setWithIsDebug:(BOOL)isDebug resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(isDebug, isDebugWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setTestDeviceId, setWithTestDeviceId:(NSString *)testDeviceId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(testDeviceId, testDeviceIdWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setCuid, setWithCuid:(NSString *)cuid resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(cuid, cuidWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setUseServerSideCallback, setWithUseServerSideCallback:(BOOL)useServerSideCallback resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(useServerSideCallback, useServerSideCallbackWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)

@end

_Pragma("clang diagnostic pop")

File: ios/RNBidmadCommonModule.swift
//
//  BidmadCommon.swift
//  react-native-bidmad-plugin-test
//
//  Created by 플랫폼-오승섭-맥북 on 2023/11/30.

//

import Foundation
import OpenBiddingHelper

@objcMembers
@objc(RNBidmadCommonModule)
class RNBidmadCommonModule: NSObject, RCTBridgeModule {

    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    static func moduleName() -> String! {
        return "RNBidmadCommonModule"
    }

    func initializeSdk(appDomain: String,
                       resolver resolve: @escaping RCTPromiseResolveBlock,
                       rejecter reject: @escaping RCTPromiseRejectBlock) {
      BIDMADSetting.sharedInstance().initializeSdk(withDomain: appDomain, platform: "reactnative") { success in
        resolve(success)
      }
    }

    func reqAdTrackingAuthorization(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        BidmadCommon.reqAdTrackingAuthorization { status in
            resolve(status.rawValue)
        }
    }

    func setAdvertiserTracking(enable: Bool, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setAdvertiserTrackingEnabled(enable)
        resolve(nil);
    }

    func getAdvertiserTracking(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(BidmadCommon.getAdvertiserTrackingEnabled())
    }

    func set(isChildDirectedAds: Bool, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setIsChildDirectedAds(isChildDirectedAds)
        resolve(nil);
    }

    func isChildDirectedTreatment(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard let isChildDirectedAds = BidmadCommon.isChildDirectedTreatment() else {
            resolve(nil)
            return
        }

        resolve(isChildDirectedAds.boolValue)
    }

    func set(userConsentStatusForCCPACompliance: Bool, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setUserConsentStatusForCCPACompliance(userConsentStatusForCCPACompliance)
        resolve(nil)
    }

    func isUserConsentCCPA(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard let isUserConsentCCPA = BidmadCommon.isUserConsentCCPA() else {
            resolve(nil)
            return
        }

        resolve(isUserConsentCCPA.boolValue)
    }

    func set(isDebug: Bool, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setIsDebug(isDebug)
        resolve(nil)
    }

    func isDebug(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(BidmadCommon.isDebug())
    }

    func set(testDeviceId: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setTestDeviceId(testDeviceId)
        resolve(nil)
    }

    func testDeviceId(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(BidmadCommon.testDeviceId())
    }

    func set(cuid: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setCuid(cuid)
        resolve(nil)
    }

    func cuid(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(BidmadCommon.cuid())
    }

    func set(useServerSideCallback: Bool, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        BidmadCommon.setUseServerSideCallback(useServerSideCallback)
        resolve(nil)
    }

    func useServerSideCallback(resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(BidmadCommon.useServerSideCallback())
    }
}
